src/analysis/vector/qgszonalstatistics.cpp

```diff
@@ -439,59 +430,61 @@ void QgsZonalStatistics::statisticsFromMiddlePointTest( const QgsGeometry &poly,
   double hCellSizeY = cellSizeY / 2.0;
   double pixelArea = cellSizeX * cellSizeY;
   double weight = 0;
 
-  QgsRectangle featureBBox = poly.boundingBox().intersect( &rasterBBox );
-  QgsRectangle intersectBBox = rasterBBox.intersect( &featureBBox );
+  std::unique_ptr< QgsGeometryEngine > polyEngine( QgsGeometry::createGeometryEngine( poly.constGet( ) ) );
+  if ( !polyEngine )
+  {
+    return;
+  }
+  polyEngine->prepareGeometry();
 
-  QgsRasterBlock *block = mRasterProvider->block( mRasterBand, intersectBBox, nCellsX, nCellsY );
+  std::unique_ptr< QgsRasterBlock > block( mRasterInterface->block( mRasterBand, rasterBBox, nCellsX, nCellsY ) );
   for ( int i = 0; i < nCellsY; ++i )
   {
-    double currentX = rasterBBox.xMinimum() + cellSizeX / 2.0 + pixelOffsetX * cellSizeX;
+    double currentX = rasterBBox.xMinimum() + cellSizeX / 2.0;
     for ( int j = 0; j < nCellsX; ++j )
     {
-      if ( !validPixel( block->value( i, j ) ) )
-      {
-        continue;
-      }
-
-      pixelRectGeometry = QgsGeometry::fromRect( QgsRectangle( currentX - hCellSizeX, currentY - hCellSizeY, currentX + hCellSizeX, currentY + hCellSizeY ) );
-      if ( !pixelRectGeometry.isNull() )
+      double pixelValue = block->value( i, j );
+      if ( validPixel( pixelValue ) && !block->isNoData( i, j ) )
       {
-        //intersection
-        QgsGeometry intersectGeometry = pixelRectGeometry.intersection( poly );
-        if ( !intersectGeometry.isNull() )
+        pixelRectGeometry = QgsGeometry::fromRect( QgsRectangle( currentX - hCellSizeX, currentY - hCellSizeY, currentX + hCellSizeX, currentY + hCellSizeY ) );
+        // GEOS intersects tests on prepared geometry is MAGNITUDES faster than calculating the intersection itself,
+        // so we first test to see if there IS an intersection before doing the actual calculation
+        if ( !pixelRectGeometry.isNull() && polyEngine->intersects( pixelRectGeometry.constGet() ) )
         {
-          double intersectionArea = intersectGeometry.area();
-          if ( intersectionArea >= 0.0 )
+          //intersection
+          QgsGeometry intersectGeometry = pixelRectGeometry.intersection( poly );
+          if ( !intersectGeometry.isEmpty() )
           {
-            weight = intersectionArea / pixelArea;
-            stats.addValue( block->value( i, j ), weight );
+            double intersectionArea = intersectGeometry.area();
+            if ( intersectionArea > 0.0 )
+            {
+              weight = intersectionArea / pixelArea;
+              stats.addValue( pixelValue, weight );
+            }
           }
         }
-        pixelRectGeometry = QgsGeometry();
       }
       currentX += cellSizeX;
     }
     currentY -= cellSizeY;
   }
-  delete block;
 }
 
-bool QgsZonalStatistics::validPixel( float value ) const
+bool QgsZonalStatistics::validPixel( double value ) const
 {
-  return !( value == mInputNodataValue || std::isnan( value ) );
+  return !std::isnan( value );
```

Comment:
Do we really need this validPixel function here?

Reply:
Do you mean "do we need the check" or "does this have to be a function"?

Reply:
@nyalldawson , "does this have to be a function?" :smile:

Reply:
Not sure... I'm thinking there's *probably* an issue if non-finite pixel values are present here which should also be caught by the function, but I'm reluctant to add this extra check without knowing that there is actually an issue. So I'd keep it like this to make it easy to possibly refine like this in future....

Reply:
(it's also called in more than one place)

Reply:
OK